### PR TITLE
[inductor] Fuse a q-by-kv contraction in flex_attention score_mod into a tl.dot

### LINKED ISSUE #148112: MLA with Learnable RoPE Tensors is Broken with Flex Attention
### 🐛 Describe the bug

I'm encountering an `AssertionError` during the backward pass when using `flex_attention` if the tensors used require gradients.

```python
import torch
from torch import Tensor
from torch.nn.attention.flex_attention import flex_attention

torch._inductor.config.unroll_reductions_threshold = 65

def generate_mla_rope_score_mod(query_rope: Tensor, key_rope: Tensor, num_heads: int, scale: float = 1.0):
    def mla_rope_score_mod(score: Tensor, b: Tensor, h: Tensor, q_idx: Tensor, kv_idx: Tensor) -> Tensor:
        return score + (scale * torch.dot(query_rope[b, h, q_idx], key_rope[b, h // num_heads, kv_idx]))

    return mla_rope_score_mod

def main(device: str = "cuda"):
    B, H, SEQ_LEN, LATENT_HEAD_DIM, ROPE_HEAD_DIM = 1, 128, 8, 512, 64

    query = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)
    key = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)
    value = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)

    query_pe = torch.rand(B, H, SEQ_LEN, ROPE_HEAD_DIM, device=device, requires_grad=True)
    key_pe = torch.rand(B, H, SEQ_LEN, ROPE_HEAD_DIM, device=device, requires_grad=True)

    score_mod = generate_mla_rope_score_mod(query_pe, key_pe, ROPE_HEAD_DIM)
    out = torch.compile(flex_attention)(query, key, value, score_mod=score_mod)
    out.sum().backward()

main()
```

```
AssertionError:
  File "/home/ubuntu/.local/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 3440, in fn
    assert len(idx) == len(output_size)
```

The full traceback shows that it's happening in the `flex_attention_backward` function within the inductor's kernel module.

I can confirm it works if you don't require gradients for `query_pe` and `key_pe`, but it seems like it should work if you do (and I have an adjacent application where they need to require gradients).

### Versions

'2.7.0a0+gitce805a5'

cc @chauhang @penguinwu @ydwu4 @bdhirsh @bobrenjc93 @aorenste @Chillee @drisspg @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv @zou3519

### PR DESCRIPTION
**Pull request by Ansh-Karnwal**
### What

An MLA-style `score_mod` contracts a query-indexed vector against a key-indexed one:
```python
score + torch.dot(query_rope[b, h, q_idx], key_rope[b, h, kv_idx])
```
This PR recognizes that term in the `score_mod` subgraph and computes it in the template with one extra `tl.dot` on the score tile, instead of unrolling it into R multiply-adds per score element. Forward only, gated by `torch._inductor.config.flex_fuse_score_mod_contraction` (default on).

### Why

`score_mod` subgraphs are pointwise by construction: the renderer evaluates them at the empty index and all tile-ness comes from Triton broadcasting the inputs, so a reduction must be realized into a buffer, which `PointwiseSubgraphLowering.register_buffer` refuses. Two workarounds exist today, and both are bad:

1. Raise `unroll_reductions_threshold` above R so the dot expands into R pointwise multiply-adds per element of the `[BLOCK_M, BLOCK_N]` tile. This is what `attention-gym`'s MLA example does.
2. Precompute `query_rope @ key_rope.T` into a `[B, H, S, S]` score bias, which costs O(S^2) memory and defeats the point of FlexAttention.

The memory argument needs no stopwatch, and it is the one that matters: the bias is O(S^2) while the operands are O(S*R), so at real MLA context lengths option 2 stops fitting and option 1 is what you are left with. See the memory columns below.

The key point is that this contraction **factorizes**. The `[BLOCK_M, BLOCK_N, R]` intermediate the unrolled form materializes is exactly `A_tile[BLOCK_M, R] @ B_tile[R, BLOCK_N]`, so the template can compute it with one `tl.dot` and never form the rank-3 intermediate - which is how FlashMLA does it. Generic reductions in `score_mod` (#141627) are infeasible as framed (at 128x128x64 the intermediate is a million elements), but this subset is tractable and is the one real workloads want.

Measured on an RTX 3080 (sm_86), B=1 H=8 D=128 R=64 fp16, one fresh process per measurement, median of 5 batches x 20 reps, best over `{default, 64x64, 128x64}` block sizes, with a profiler check that a Triton flex kernel actually ran and a checksum anchor:

| | time S=1024 | time S=4096 | extra memory S=4096 | extra memory S=8192 |
|---|---|---|---|---|
| no rope term at all (floor) | 0.152 ms | 1.420 ms | - | - |
| hoisted `[B, H, S, S]` bias | 0.149 ms | 1.807 ms | 256 MiB | 1 GiB |
| unrolled dot in `score_mod` | 0.649 ms | 9.032 ms | 8 MiB | 16 MiB |
| **fused (this PR)** | **0.150 ms** | **1.79 ms** | **8 MiB (32x less)** | **16 MiB (64x less)** |

The two memory columns are allocation sizes - exact arithmetic, not measurements - and count only what the rope term adds on top of q/k/v. S=8192 was not benchmarked for time; it is in the table because the quadratic-versus-linear divergence is the whole argument. The S=4096 fused time is the median of three separate runs (1.74 / 1.79 / 1.87); run-to-run spread on this box is around 5%, so treat differences smaller than that as noise.

So the fused form is **4.3x faster at S=1024 and 5.0x faster at S=4096** than the unrolled path. Against the hoisted bias it is level on time and uses 32x less extra memory at S=4096, widening to 64x at S=8192. At S=1024 it lands on the no-rope floor - the contraction is essentially free - and at S=4096 it sits about 25% above the floor, because the operand tiles are re-read over a kv loop that is 4x longer.

That last point invites an obvious optimization: the q-side tile is loop invariant, so it could be loaded once in the kernel body and passed into the block loop. I implemented that and measured it as a wash (1.78 / 1.89 against 1.74 / 1.79 / 1.87), presumably trading the reload, which hits cache, for register pressure. So this PR keeps the simpler version with both loads inside the block function.

### Design question

The part I would most like guidance on: **should this be pattern recognition on `score_mod`, or a first-class API?**

This PR recognizes a contraction out of the graph the user already writes, so existing MLA code gets faster with no API change and no `unroll_reductions_threshold` hack. The alternative is to expose an explicit mod point - a `qk_mod` or an MLA-shaped argument - and have the template contract declared operands. That is a bigger change and a user-visible API, but it is honest about what the kernel can do instead of relying on the compiler to spot a shape.

The forward is the shared prefix of either design, which is why it is what this PR contains.

### Scope

Not included, deliberately:

- **The backward, so this does not close #148112.** Learnable rope needs gradients for the captured operands, which arrive as `flex_lib.zeros_and_scatter` calls that are partially indexed with vector payloads - the case that trips `select_algorithm.py`'s scatter store. The joint graph also reaches the Inductor lowering already traced, so a forward-side rewrite cannot touch it. A fused backward needs its own joint-graph pattern (`dqr = ds @ kr_tile`, `dkr = ds^T @ qr_tile`; the math does factorize) plus that scatter fix.
- **Decode.** A decode tile spans several query heads, so the kv-side operand differs per row of the tile and the term is not a single matmul there. Decode shapes keep today's behaviour; there is a test for that.
- Generic reductions in `score_mod`/`mask_mod`, transformed head indices (GQA's `h // n`), more than one contraction per `score_mod`, and a symbolic contracted extent. All bail to today's path. A dynamic sequence length is fine and stays fused; it is only a symbolic *extent* that bails, since the tile shape has to be a compile-time constant, so `dynamic=True` (which marks every dim) turns the fusion off.

One numerics note: the contraction dot uses `input_precision="ieee"` rather than the kernel's `FLOAT32_PRECISION`. `score_mod` spelled this as a reduction, not a matmul, and letting float32 operands fall to tf32 made the output 3.4x and the gradients up to 28x less accurate than the unrolled form.

### Testing

New tests in `test/inductor/test_flex_attention.py`: numerics against the float64 golden for R in {16, 17} x {full, causal} including gradients; fused-versus-unrolled A/B for R in {8, 48, 64, 128} and both spellings of the dot; a codegen check that exactly one extra `tl.dot` appears; five score_mods the detector must reject staying unfused; the `max_autotune` path (the one that materializes benchmark tensors, so it would catch a wrong operand order) with a decoy capture ahead of the two operands; decode shapes falling back; and the backward being unchanged with the flag on.

Verified separately on the same box: 22 shape/dtype combinations against an einsum reference (fp16/bf16/fp32, R from 8 to 128 including non-powers of two, non-divisible sequence lengths, Sq != Skv, GQA, H=1, a sliced-view capture with a nonzero storage offset, and a dynamic sequence length), and byte-identical generated code with the flag on and off for every score_mod the detector rejects.

Not validated locally: sm_86 has 99 KiB of shared memory, so large-R with large blocks (H100 territory) is untested here. Leaving that, and the full flex suites on other backends, to CI.

### CODE REVIEW

**Comment by drisspg** [comment]
while this is quite cool ;  the fact that MLA worked at all was more quirk and not intended; this needs more design. Feel free to open an issue and we can talk over there
